charmhelpers/contrib/openstack/context.py

```diff
@@ -2244,3 +2248,119 @@ def __call__(self):
+
+    # Everything is OK
+    return None, None
+
+
+class DHCPAgentContext(OSContextGenerator):
+
+    def __call__(self):
+        """Return the 'default_availability_zone' from the principal that this
+        ovs unit is attached to (as a subordinate) and the 'dns_domain' from
+        the neutron-plugin-api relations (if one is set).
+
+        :returns: {} if no relation set, or
+            {'availability_zone': availability_zone from principal relation}
+        """
+
+        ctxt = {}
+        dnsmasq_flags = config('dnsmasq-flags')
+        if dnsmasq_flags:
+            ctxt['dnsmasq_flags'] = config_flags_parser(dnsmasq_flags)
+        ctxt['dns_servers'] = config('dns-servers')
+
+        neutron_api_settings = NeutronAPIContext()()
+
+        ctxt['debug'] = config('debug')
+        ctxt['instance_mtu'] = config('instance-mtu')
+        ctxt['ovs_use_veth'] = self.get_ovs_use_veth()
+
+        ctxt['enable_metadata_network'] = config('enable-metadata-network')
+        ctxt['enable_isolated_metadata'] = config('enable-isolated-metadata')
+
+        if neutron_api_settings.get('dns_domain'):
+            ctxt['dns_domain'] = neutron_api_settings.get('dns_domain')
+
+        # Override user supplied config for these plugins as these settings are
+        # mandatory
+        if config('plugin') in ['nvp', 'nsx', 'n1kv']:
+            ctxt['enable_metadata_network'] = True
+            ctxt['enable_isolated_metadata'] = True
+
+        return ctxt
+
+    def get_existing_ovs_use_veth(self):
+        """Return existing ovs_use_veth setting from dhcp_agent.ini.
+
+        :returns: Value of ovs_use_veth setting: None or Bool
+        :rtype: Union[None, Bool]
+        """
+        dhcp_agent_ini = "/etc/neutron/dhcp_agent.ini"
+        existing_ovs_use_veth = None
+        # If there is a dhcp_agent.ini file read the current setting
+        if os.path.isfile(dhcp_agent_ini):
+            # config_ini does the right thing and returns None if the setting is
+            # commented.
+            existing_ovs_use_veth = (
+                config_ini(dhcp_agent_ini)["DEFAULT"].get("ovs_use_veth"))
+        return existing_ovs_use_veth
+
+    def get_ovs_use_veth(self):
+        """Return correct ovs_use_veth setting.
+
+        Get the right value from config or existing dhcp_agent.ini file.
+        Existing has precedence. Attempt to default to "False" without
+        disrupting existing deployments. Handle existing deployments and
+        upgrades safely. See LP Bug#1831935
+
+        :returns: Value to use for ovs_use_veth setting
+        :rtype: string
+        """
+        _existing = self.get_existing_ovs_use_veth()
+        if _existing is not None:
+            return str(_existing)
+
+        _config = config('ovs-use-veth')
+        if _config is None:
+            # New better default
+            return "False"
+        else:
+            return str(_config)
```

Comment:
I think this needs more validation of the input.  It's a String, but really it's a trinary boolean value of True, False, None.

I think (up for discussion) that this function should return Boolean (True/False) and do the work in the function to extract the boolean from the str.  i.e. parse it to a type here and then that type is the version to use.  My rationale is that if a str is allowed to escape this function without validation, ovs-use-veth could be set to "None" as a string, not realising that it should really be a boolean.

Reply:
Agreed. Adding a parse_ovs_use_veth method to convert string to bool and then this method will return bool.